File: from-1201-to-1300/1239-maximum-length-of-a-concatenated-string-with-unique-characters/solution.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        char_set = set()
        
        def hasDuplicates(cs: set, s: string)-> bool:
            pre_set = set()
            for c in s:
                if c in pre_set or c in cs:
                    return True
                pre_set.add(c)
            return False
        
        def backtrack(i: int):
            if i == len(arr):
                return len(char_set)
            
            ans = 0
            if not hasDuplicates(char_set, arr[i]):
                for c in arr[i]:
                    char_set.add(c)
                ans = backtrack(i+1)
                for c in arr[i]:
                    char_set.remove(c)
                    
            return max(ans, backtrack(i+1))
        
        return backtrack(0)

File: from-1201-to-1300/1239-maximum-length-of-a-concatenated-string-with-unique-characters/solution.go
func maxLength(arr []string) int {
    currSet := make(map[rune]bool)
    canAdd := func(currSet map[rune]bool, candidate string) bool {
        candSet := make(map[rune]bool)
        for _, r := range candidate {
            if candSet[r] || currSet[r] {
                return false
            }
            candSet[r] = true
        }
        return true
    }
    var backtrack func(i int) int
    backtrack = func(i int) int {
        if i == len(arr) {
            return len(currSet)
        }
        var ans int = 0
        if canAdd(currSet, arr[i]) {
            for _, r := range arr[i] {
                currSet[r] = true
            }
            ans = backtrack(i + 1)
            for _, r := range arr[i] {
                delete(currSet, r)
            }
        }
        a2 := backtrack(i + 1)
        if a2 > ans {
            return a2
        }
        return ans
    }
    return backtrack(0)
}